File: functions/device/get_device.py
from datetime import datetime

from firebase_admin import firestore
from flask import Request, abort
from models.constants import Collections, DeviceFields
from models.data_models import Device
from utils.logging_utils import get_logger


def get_device(request: Request):
    """
    Get the device information for the authenticated user.

    Args:
        request: The Flask request object with user_id attached from authentication.

    Returns:
        A Device object containing the device information if found.

    Raises:
        404: Device not found
    """
    logger = get_logger(__name__)
    logger.info(f"Getting device for user {request.user_id}")

    # Reference to the user's device document
    db = firestore.client()
    device_ref = db.collection(Collections.DEVICES).document(request.user_id)

    # Get the device document
    device_doc = device_ref.get()
    if not device_doc.exists:
        logger.warning(f"Device not found for user {request.user_id}")
        abort(404, description="Device not found")

    # Get device data and create Device object
    device_data = device_doc.to_dict()
    logger.info(f"Device retrieved for user {request.user_id}")

    updated_at = device_data.get(DeviceFields.UPDATED_AT, "")
    if isinstance(updated_at, datetime):
        updated_at = updated_at.isoformat() + "Z"

    return Device(
        device_id=device_data.get(DeviceFields.DEVICE_ID, ""), updated_at=updated_at
    )

File: functions/src/device/get-device.ts
import { Request, Response } from 'express';
import { getFirestore, Timestamp } from 'firebase-admin/firestore';
import { Collections, DeviceFields } from '../models/constants.js';
import { Device } from '../models/data-models.js';
import { NotFoundError } from '../utils/errors.js';
import { getLogger } from '../utils/logging-utils.js';
import { formatTimestamp } from '../utils/timestamp-utils.js';

import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const logger = getLogger(path.basename(__filename));

/**
 * Get the device information for the authenticated user.
 *
 * @param req - The Express request object containing:
 *              - userId: The authenticated user's ID (attached by authentication middleware)
 * @param res - The Express response object
 *
 * @returns A Device object containing the device information if found
 *
 * @throws {404} Device not found
 */
export const getDevice = async (req: Request, res: Response): Promise<void> => {
  const currentUserId = req.userId;
  logger.info(`Getting device for user ${currentUserId}`);

  // Reference to the user's device document
  const db = getFirestore();
  const deviceRef = db.collection(Collections.DEVICES).doc(currentUserId);

  // Get the device document
  const deviceDoc = await deviceRef.get();
  if (!deviceDoc.exists) {
    logger.warn(`Device not found for user ${currentUserId}`);
    throw new NotFoundError('Device not found');
  }

  // Get device data and create Device object
  const deviceData = deviceDoc.data();
  logger.info(`Device retrieved for user ${currentUserId}`);

  // Format timestamp for consistent API response
  const updatedAt = deviceData?.[DeviceFields.UPDATED_AT] as Timestamp;
  const updatedAtIso = updatedAt ? formatTimestamp(updatedAt) : '';

  const device: Device = {
    device_id: deviceData?.[DeviceFields.DEVICE_ID] || '',
    updated_at: updatedAtIso,
  };

  res.json(device);
};
